### PR TITLE
Never enable JSX in a .ts file

### LINKED ISSUE #6849: Legacy TS cast: Expected jsx identifier
I encountered an issue that is probably related to this one, using `^2.0.0-nightly.734`, in a `.ts` file, casting using the `<Type>` syntax doesn't work.

```ts
  const setAction = (fields: Partial<HeaderAction>) =>
    setCurrentAction({ ...<HeaderAction>getCurrentAction(), ...fields })
```
    
throws 
```
Building...
× Build failed.
@parcel/transformer-js: Unexpected token `HeaderAction`. Expected jsx identifier
C:\dev\oz\web\src\api\header\actions.ts:36:28
  35 |   const setAction = (fields: Partial<HeaderAction>) =>
> 36 |  ...<HeaderAction>getCurrentAction(), ...fields 
})
>    |      ^^^^^^^^^^^^
  37 |
  38 |   _actions([
```

This started happening once i removed my .babelrc which contained
```json
{
  "presets": [
    "@babel/preset-typescript",
    "@babel/preset-react"
  ],
  "plugins": [
    "@babel/plugin-proposal-throw-expressions"
  ]
}
```
after trying to upgrade to a newer Parcel version where it threw a warning saying that my presets were already Parcel's default.

_Originally posted by @Banou26 in https://github.com/parcel-bundler/parcel/issues/6719#issuecomment-907557726_

### PR DESCRIPTION
**Pull request by devongovett**
Fixes  #6849

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.19s | -27.00ms |
| Cached | 326.00ms | +21.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 648.00ms | -38.00ms 🚀 |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 649.00ms | -37.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 649.00ms | -37.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 81.00ms | -615.00ms 🚀 |
| dist/legacy/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 82.00ms | -615.00ms 🚀 |
| dist/modern/parcel.7cdb0fad.webp | 102.94kb | +0.00b | 82.00ms | -615.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.76s | +164.00ms |
| Cached | 496.00ms | +18.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.20m | +110.00ms |
| Cached | 1.74s | -53.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/EmojiPickerComponent.4259176d.js | 146.67kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/esm.304eb046.js | 33.15kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/DatePicker.05d89a3e.js | 22.98kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/js.0f0bb621.js | 17.25kb | +0.00b | 41.38s | -12.68s 🚀 |
| dist/png-chunks-extract.94b5b9fc.js | 3.58kb | +0.00b | 41.38s | -12.68s 🚀 |
| dist/feedback.e18b45f0.js | 1.77kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/workerHasher.567cfc57.js | 1.63kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/heading6.b3b946d1.js | 1.51kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/heading5.e617db66.js | 1.38kb | +0.00b | 22.39s | -11.15s 🚀 |
| dist/expand.1f17de7f.js | 1.29kb | +0.00b | 22.39s | -11.15s 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.0b454c47.js | 1.78mb | +0.00b | 20.35s | -1.03s 🚀 |
| dist/EmojiPickerComponent.4259176d.js | 146.67kb | +0.00b | 22.26s | -12.46s 🚀 |
| dist/component.3b0cab6d.js | 37.35kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/esm.304eb046.js | 33.15kb | +0.00b | 22.27s | -12.46s 🚀 |
| dist/component.8869d70a.js | 24.94kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/DatePicker.05d89a3e.js | 22.98kb | +0.00b | 22.26s | -12.46s 🚀 |
| dist/js.0f0bb621.js | 17.25kb | +0.00b | 41.20s | -14.31s 🚀 |
| dist/png-chunks-extract.94b5b9fc.js | 3.58kb | +0.00b | 41.20s | -14.31s 🚀 |
| dist/16.7b42ca8a.js | 2.35kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/date.c0482a02.js | 1.87kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/feedback.e18b45f0.js | 1.77kb | +0.00b | 22.26s | -12.46s 🚀 |
| dist/workerHasher.567cfc57.js | 1.63kb | +0.00b | 22.27s | -12.46s 🚀 |
| dist/code.cd95a955.js | 1.51kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/heading6.b3b946d1.js | 1.51kb | +0.00b | 22.25s | -12.47s 🚀 |
| dist/heading5.e617db66.js | 1.38kb | +0.00b | 22.25s | -12.47s 🚀 |
| dist/expand.1f17de7f.js | 1.29kb | +0.00b | 22.26s | -12.46s 🚀 |
| dist/action.9e337641.js | 1.14kb | +0.00b | 20.34s | -1.03s 🚀 |
| dist/component.edb11b95.js | 1.13kb | +0.00b | 20.34s | -1.03s 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.64s | +121.00ms |
| Cached | 431.00ms | +7.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/311760957257810496)**

File: packages/transformers/js/src/JSTransformer.js [hunk 1]
```diff
@@ -222,11 +222,16 @@ export default (new Transformer({
         }
       }
 
-      isJSX = Boolean(
-        compilerOptions?.jsx ||
-          pragma ||
-          JSX_EXTENSIONS[path.extname(config.searchPath)],
-      );
+      let ext = path.extname(config.searchPath);
+      if (ext === '.ts') {
```
Comment (mischnic) [inline packages/transformers/js/src/JSTransformer.js]: Should this rather be checking `asset.type`?

Comment (devongovett) [reply packages/transformers/js/src/JSTransformer.js]: Hmm, the asset isn't passed to loadConfig... 😕

**Review by mischnic — APPROVED** [review]
I suppose this could still break with a Vue or Svelte inline typsecript asset, but at least this common case is fixed.

**Comment by devongovett** [comment]
Haha oops, nice timing. I moved the extension checking into transform instead of loadConfig.

**Review by mischnic — APPROVED** [review]

**Pull request merged**